### DISCUSSION TITLE
Response Headers

### DISCUSSION THREAD

**Discussion by kcparker** [go-kit/kit#1247] [Q&A] [answered]
Hi, 

I want to use [ETags](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) to avoid mid-air-collisions. During get requests the response should have an ETag header, so that a subsequent update request with contain that value in an ETag request header.

The value of the ETag will come from the service layer which is invoked by the Endpoint. I am using a grpc transport, and [ServeGRPC](https://github.com/go-kit/kit/blob/master/transport/grpc/server.go#L90) does not provide a way for me to pass the value back from the Endpoint response. How should I get a value from the service layer into a response header using grpc transport? Should the context be returned from Endpoint or should there be another server after middleware that takes the response from the Endpoint?

Thank you.

**Comment by peterbourgon** [comment]
AFAIK, and please correct me if I'm mistaken, [gRPC headers](https://pkg.go.dev/google.golang.org/grpc#SendHeader) are comparable to, but not the same as, HTTP headers. Is there some prior art here? I would generally expect this to be solvable via [SetResponseHeader](https://pkg.go.dev/github.com/go-kit/kit/transport/grpc#SetResponseHeader).

**Reply by dvavili** [reply]
Hi @peterbourgon , I work with @kcparker and let me expand on the situation. We have our services exposing both grpc and http endpoints with grpc-gateway functionality. Our transport layer is pretty much grpc, but external clients use http endpoints to interact with the service. We need to add Etag support to some of our endpoints. 

The issue is that Etag is pretty much coming from the service/db layer, but we are unable to pass that information back in the form of response header. The gap that we see is that server after functions that can be utilized for setting response headers has no way to read information from the response, and the server encoder functions that take in both context and response has no way of passing back a new context with details populated from the response. Let us know if we are missing any other way to make this possible.

Otherwise, we were thinking of proposing changes to either server after/encode functions or maybe add another pipeline stage like `ctx = s.responseProcessor(ctx, response)` between the after and encode functions. We've run into such issues before and have been using workarounds, but feel that adding support to modify context on the response path based on some fields in response has a lot of value-add to go-kit. Please let us know your thoughts.

**Reply by peterbourgon** [reply]
But gRPC does not give you HTTP semantics, right? It happens to be implemented on top of HTTP/2 right now, but that's an implementation detail, not a guarantee, AFAIK. I could be wrong! Is there there anything in gRPC's spec which says that a gRPC service is going to be able to express HTTP headers?

**Reply by kcparker** [reply]
go-kit already has support for setting response headers and trailers from server after middleware [here](https://github.com/go-kit/kit/blob/master/transport/grpc/server.go#L127) but it is too limited because you cannot access any data returned back from the service layer/Endpoint response from those ServerAfterFunc(s). We are just trying to suggest this is extended to allow data in the Endpoint response to go into the mdHeader or mdTrailer.

**Reply by dvavili** [reply]
gRPC in itself doesn't, but you can customize the grpc-gateway to add HTTP headers from metadata. Ref: [Mapping from grpc server metadata to http response headers](https://grpc-ecosystem.github.io/grpc-gateway/docs/mapping/customizing_your_gateway/#mapping-from-grpc-server-metadata-to-http-response-headers). This function is called for all key-value pairs in metadata. 

The proposal is that if there is a capability to add <key, value> pair to `mdHeaders` using `s.after` function, it will let grpc-gateway to send this as HTTP response headers. However, what's missing is to populate the context with the necessary details from response so that `s.after` function has this information. Hope that makes sense.

**Reply by peterbourgon** [reply]
[ServerResponseFunc](https://pkg.go.dev/github.com/go-kit/kit/transport/grpc#ServerResponseFunc) already gives you mutable access to the md.Headers, doesn't it? Is this not sufficient?

**Reply by dvavili** [reply]
No, we hoped that was sufficient. But we are not able to get the information needed. Say, in this particular case, Etag is fetched from DB - there is no way for this information to be propagated into the context. `ServerResponseFunc` doesn't seem to have access to the response object, and there is nothing before the `ServerResponseFunc` stage to allow us to populate the context. That's the ask here.

Basically, something like 
```
--- a/transport/grpc/server.go
+++ b/transport/grpc/server.go
@@ -24,6 +24,7 @@ type Server struct {
        dec          DecodeRequestFunc
        enc          EncodeResponseFunc
        before       []ServerRequestFunc
+       rp           []ServerResponseProcessorFunc
        after        []ServerResponseFunc
        finalizer    []ServerFinalizerFunc
        errorHandler transport.ErrorHandler
@@ -61,6 +62,13 @@ func ServerBefore(before ...ServerRequestFunc) ServerOption {
        return func(s *Server) { s.before = append(s.before, before...) }
 }

+// ServerResponseProcessor functions are executed to parse information from response object
+// context or error and adds information to context.
+// This is useful for send back information to client from response.
+func ServerResponseProcessor(rp ...ServerResponseProcessorFunc) ServerOption {
+       return func(s *Server) { s.rp = append(s.rp, rp...) }
+}
+
 // ServerAfter functions are executed on the gRPC response writer after the
 // endpoint is invoked, but before anything is written to the client.
 func ServerAfter(after ...ServerResponseFunc) ServerOption {
@@ -124,6 +132,10 @@ func (s Server) ServeGRPC(ctx context.Context, req interface{}) (retctx context.
                return ctx, nil, err
        }

+       for _, f := range s.rp {
+               ctx = f(ctx, response, err)
+       }
+
        var mdHeader, mdTrailer metadata.MD
        for _, f := range s.after {
                ctx = f(ctx, &mdHeader, &mdTrailer
```
Or something equivalent that allows response and context to be present in the same function - so context can be modified based on response.

**Reply by peterbourgon** [reply]
>  But we are not able to get the information needed. Say, in this particular case, Etag is fetched from DB - there is no way for this information to be propagated into the context

Sure there is! You just need to use a pattern like e.g. [ctxdata](https://github.com/peterbourgon/ctxdata). Would that work?

**Reply by kcparker** [reply]
@peterbourgon the problem is the ctx passed into the service layer is immutable and would not have data in it that is added from the service layer by the time the ServerResponseFunc looked at the ctx.

Here is a look at the code in question.

```
response, err = s.e(ctx, request)
if err != nil {
	s.errorHandler.Handle(ctx, err)
	return ctx, nil, err
}

var mdHeader, mdTrailer metadata.MD
for _, f := range s.after {
	ctx = f(ctx, &mdHeader, &mdTrailer)
}
```

`s.e(ctx, request)` is where we would need to return the value for our header. We could return that value in the `response` but the `response` does not get passed to the `s.after` functions. The `s.after` functions get passed the `ctx` but `s.e` cannot modify that `ctx`.

**Reply by peterbourgon** [reply]
>  the problem is the ctx passed into the service layer is immutable and would not have data in it that is added from the service layer by the time the ServerResponseFunc looked at the ctx.

The ctxdata package I linked solves this problem. You inject a mutable value to the context at the top of the stack, write to that value deeper in the stack, and then read the written data on the way back out.

**Reply by dvavili** [reply]
Thanks @peterbourgon. Will give this a try and get back with results.

**Reply by dvavili** [reply]
Just wanted to add a note that I think it's cool how `ctxdata` solves for the immutable nature of context by using channels. 👍

**Reply by kcparker** [reply]
Thank you @peterbourgon this does solve our our problem.

**Comment by peterbourgon** [comment]
It's also worth noting that ETags by themselves don't solve this (or any) problem, it's only when they're combined with a caching middlebox that knows how to parse them that they have any effect. And it is IME rare that ETags make sense for Go kit applications — the caching/versioning semantics of HTTP are not usually a good match for the requirements of typical microservices.

**Reply by dvavili** [reply]
Like explained in the above comment, populating Etags is an issue that we are facing currently. Our situation might be a different than the norm due to the nature of our setup, but we've run into similar issues before with some features we've been trying to add. Basically add any information back to response headers based on response from service layer. This could be a valid case for someone utilizing grpc-gateway functionality along with go-kit grpc transport.

**Comment by basvanbeek** [comment, marked as answer]
Not directly visible here unless expanding the thread above is the actual answer how to solve this problem by @peterbourgon:

https://github.com/go-kit/kit/discussions/1247#discussioncomment-3421873

**Reply by peterbourgon** [reply]
Nice! 

Contexts are tricky. When I figured out [ctxdata](https://github.com/peterbourgon/ctxdata) it unlocked a lot of capabilities 👍
